Modules/+Sources/Msquared.m

```diff
@@ -0,0 +1,481 @@
+
+                    if obj.VIS_wavelength > 0 && obj.NIR_wavelength > 0
+                        diff_wavelength_ = 1/(1/obj.VIS_wavelength - 1/obj.NIR_wavelength);
+
+                        if diff_wavelength_ < 1951 && diff_wavelength_ > 1949   % This range is just so a silly value isn't set. Not sure if this is a good range.
+                            obj.diff_wavelength = diff_wavelength_;
+                        end
+                    end
+                else
+                    obj.hwserver.com('wavemeter', 'SetSwitcherSignalStates', obj.VIS_channel, 0, 0);
+                    obj.VIS_wavelength = NaN;
+                end
+            end
+        end
+    end
+    
+    methods     % The meat of this tunable source.
+        function tune(obj, target)                  % This is the tuning method that interacts with hardware (target in nm)
+            if isnan(target); return; end           % Do nothing if NaN
+            
+            assert(~isempty(obj.hwserver) && isobject(obj.hwserver) && isvalid(obj.hwserver), 'No hwserver host!')
+            
+            module = obj.determineModule(target);
+            
+            if isempty(module)
+                error(['Wavelength ' num2str(target) ' nm is outside the range of this msquared laser.'])
+            end
+            
+            obj.updatingVal = true;
+            obj.active_module = module;
+
+            if strcmp(module, obj.moduleVIS)
+                obj.emm_setpoint = target;
+                try
+                    out = obj.com('set_wavelength', 'EMM', target, .1); % last arg is timeout
+                
+                    if out.status ~= 0
+                        error('Failed to set EMM target')
+                    end
+                catch
+                    % expected?
+                end
+            else
+                obj.emm_setpoint = NaN;
+            end
+            
+            attempting = true;
+            failcount = 0;
+            
+            while attempting
+                nir_wavelength = obj.determineNIRWavelength(target);
+                obj.solstis_setpoint = nir_wavelength;
+                out = obj.com('set_wavelength', 'solstis', nir_wavelength, 0); % last arg is timeout
+                
+                if out.status ~= 0  % Call failed
+                    failcount = failcount + 1;
+                else
+                    obj.tuning = true;
+
+                    obj.getFrequency();
+
+                    obj.trackFrequency(obj.c/target);   % Will block until obj.tuning = false (calling obj.getFrequency each tick)
+                
+                    if ~obj.center_percent || (obj.resonator_voltage < 120 && obj.resonator_voltage > 80)    % If we are good with anything or we are inside the acceptable range...
+                        attempting = false;
+                    else
+                        detuning = 1 - 2*(nir_wavelength > 900);    % Tune up if wavelength < 900, down otherwise.
```

Comment:
is this a reasonable range to detune? 

also, should this process have a timeout-->error?

Reply:
1 nm is far enough to get it to coarse tune, I think, which is the goal.